feat(studio): scoped personal access tokens ui

kemaldotearth opened the pull request:
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES

## What kind of change does this PR introduce?

This PR introduces a more advanced UI for the scoped personal access tokens found in the users account area. This PR introduces.

## What is the current behavior?

The current behaviour is limited in the sense that users cannot set things like token expiry or granular permissions. 

## What is the new behavior?

This PR aims to bring in the kind of control our access tokens previously lacked. In this we address the following:

- Tidy up of the Access Token area. This is more inline with future UI direction.
- An improved Create New Token flow.
- Ability to configure an Expiry date for access tokens.
- Ability to scope your access token's by organizations or projects.
- Ability to set granular access for a variety of permissions.
- Once a token has been created and configured, we also have a panel to show you how everything is configured.

## Additional context

🚧 Awaiting API endpoint updates to be able to get this in there.

Commit:
feat: introduce manage tokens entry point

This adds a button to trigger the eventual manage tokens tray

Commit:
feat: add delete token into more dropdown

To add a bit consistency with other table rows, we put the delete inside a more dropdown. Also useful if more general actions will come for tokens.

Commit:
feat: add manage tokens panel

This adds the sheet where the PATs will be managed.

Commit:
fix: add truncation to access token name column

This adds truncation the access token name as currently if you set a really long name it distorts the table.

Commit:
chore: add some dummy data for now

Adds some dummy data to play with while I prototype.

Commit:
chore: slightly adjusted direction

As we can only configure permissions on create, this introduces beginnings for project access and permission settings in the create dialog. This also changes the update panel to a view only panel.

Commit:
feat: step by step concept

Commit:
feat: access token view layout feedback

Implements the feedback from Saxon on the page layout.

Commit:
feat: alternative no wizard option

This is a version where everything is inside the dialog without steps to guide you through. It is much more upfront.

Commit:
feat: side panel concept

Commit:
chore: tidying up some of cursors files

Cursor leaves behind unused files, cleaning up and cleaning up the final prototype in general.

Commit:
fix: bring old admonition back for copy token

This brings back the old admonition for copying of the token.

Commit:
chore: some folder organising

Move all access token stuff to its own folder as its got quite a few files now

Commit:
chore: remove unused token files

This removes some AI slop from my Cursor prototype. There will be a few more of these commits cleaning up.

Commit:
style: small padding update on token form

Added separators so needed to adjust the padding.

Commit:
feat: add custom date field for token expiry

Re-introduce the custom date field picker for custom token expiry dates

Commit:
style: another small padding adjustment

More in line with other sheet panels.

Commit:
feat: using the users projects and orgs for selector

This removes dummy data from resource selection and uses the orgs or projects the user has access to in the multi selector

Commit:
chore: remove unused deps

kemaldotearth committed:
chore: remove more unused things

kemaldotearth committed:
fix: error state for name field validation

Cursor did Cursor things, reverted this back to intended validation style.

kemaldotearth committed:
fix: update test file for new access token button

This updates the test file for the ccess token button.

kemaldotearth committed:
style: some token permissions ui clean up

Cleans up token permissions part of the form and fixes the functionality of the popover search

kemaldotearth committed:
feat: add ability to delete from view permissions

From the view panel after token creation, you can also delete the tokens

kemaldotearth committed:
chore: remove unused token dialog

Remove some more remnants from prototyping.

kemaldotearth committed:
style: simplify the view access sheet

kemaldotearth committed:
feat: update sorting of permission adder

The sorting of the permissions adding has been simplified.

kemaldotearth committed:
chore: remove unused constants from tokens

Removes two unused constants from our tokens section.

kemaldotearth committed:
fix: some small fixes for custom date field

This fixes the custom date fields formatting for date once you select it as well as adding a missing enum from zod

kemaldotearth committed:
fix: form tests for generate

All the tests should be passing now, fingers crossed

kemaldotearth committed:
fix: prettier warnings and errors

Some token files are not being formatted correctly with prettier, had to re-run manually

kemaldotearth committed:
style: little bit more love to the access panel

kemaldotearth committed:
feat: quick re-do of the view access sheet

This one is a bit cleaner to parse as a user. Now all I need is the generate token endpoint updated.

kemaldotearth commented on apps/studio/components/interfaces/Account/AccessTokens/NewAccessTokenButton.tsx:
```diff
@@ -0,0 +1,242 @@
+  organizationPermissions: z.record(z.string(), z.string()).optional(),
+  projectPermissions: z.record(z.string(), z.string()).optional(),
+  permissionRows: z.array(PermissionRowSchema).optional(),
+})
```
@hieuwu appreciate the endpoint may not look like this, so let's adjust accordingly once that's ready!

hieuwu replied:
Cool. Great work. I am using some tricks in my code to handle different links, now with this fix I can standardize my code more. Thanks for the fix. Let's see 🤞

kemaldotearth commented on apps/studio/components/interfaces/Account/AccessTokens/NewAccessTokenButton.tsx:
```diff
@@ -0,0 +1,242 @@
+      { name: values.tokenName, scope: tokenScope },
+      {
+        onSuccess: (data) => {
+          console.log('Generated token data:', data)
```
AI slop, get rid.

kemaldotearth commented on apps/studio/components/interfaces/Account/AccessTokens/TokenBasicInfoForm.tsx:
```diff
@@ -0,0 +1,133 @@
+  PopoverTrigger_Shadcn_,
+  Button,
+  Input_Shadcn_,
+  Input,
```
Ended up using this deprecated component due to our newer Input not gracefully handling the actions within the input, maybe something we should revisit in a separate ticket/thing @SaxonF ?

kemaldotearth commented on apps/studio/components/interfaces/Account/AccessTokens/ViewTokenPermissionsPanel.tsx:
```diff
@@ -0,0 +1,270 @@
+  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false)
+
+  // Dummy access settings for demonstration
+  const getDummyAccess = (resource: string) => {
```
This will be removed thankfully, once we have those permissions coming from somewhere.

kemaldotearth committed:
chore: remove console log

kemaldotearth commented:
This is still in progress as we await the tokens endpoint being updated with the [following](https://linear.app/supabase/project/scoped-pats-9029dc10a7f8/overview). I would appreciate a second pair of 👀 from anyone who might have a spare 5 minutes @supabase/dashboard. I tried to use the appropriate components in tandem with Cursor. 

cc @hieuwu

kemaldotearth committed:
fix: padding for experimental api admonition

This fixes the padding around the experimental API admonotion when you select that option to generate a token.

kemaldotearth commented:
@hieuwu as mentioned, potentially what we might need for the frontend to make the mutation. We could break this down if parts of the API still aren't ready and ship the UI in phases.

Currently:
```
created_at: string;
id: number;
name: string;
scope?: "V0";
token: string;
token_alias: string;
```

After: 
```
created_at: string;
id: number;
name: string;
scope?: "V0";
token: string;
token_alias: string;
expires_at: string;
resourceAccess: 'everything' | 'projects' | 'organizations';
selectedProjects?: project[];
selectedOrgs?: organization[];
permissions: permissions[];
```

Perhaps there's a better way you can deliver some of the stuff, but some of it is straight forward and others I'm not sure what shape to expect back. `expires_at` feels like something we can ship quickly, `resourceAccess` probably needs some kind of check based on the users role and `selectedProjects` / `selectedOrgs` could probably be one thing. They're only to be expected if a user tries to scope by project or organization.

As for the `permissions`, let me know what that looks like, we can organise as we wish on the frontend side. I'm not sure if these go as a different request or part of the same request as creating an access token, but perhaps we can break this form down and ship in phases?

kemaldotearth committed:
chore: organize folders

Move form components into their own folder so we can isolate them.

kemaldotearth committed:
chore: clean up unused deps

jordienr commented:
Few nitpicks:

1. Would be cool to have a "reset/remove all" permissions here
<img width="1286" height="1570" alt="CleanShot 2025-08-05 at 14 11 12@2x" src="https://github.com/user-attachments/assets/e3503c4a-16ee-46ac-ac5f-16569989640a" />

2. Missing some "read-only" presets or a way to make all permissions readonly instead of going one by one. 

3. Could add a success toast once the token is created (common pattern throughout the dashboard) 

4. Error handling (network errors or possible api errors) 
<img width="1740" height="1696" alt="CleanShot 2025-08-05 at 14 15 43@2x" src="https://github.com/user-attachments/assets/b2b9b705-44b5-4da0-9128-23b3063ca3a3" />

5. some permissions default to a value but some dont, could fix by defaulting to readonly 

<img width="1344" height="774" alt="CleanShot 2025-08-05 at 14 17 15@2x" src="https://github.com/user-attachments/assets/2578bc3a-ab09-4cbf-9128-3151c643a0a6" />

kemaldotearth committed:
feat: add reset for all permissions.

Good catch Jordi. I added a button that can reset all your permissions if you decide you need to clear them all.

kemaldotearth committed:
fix: adding individual permission selection

Adding inidividual permissions wasnt selecting the default access type, it does now.

kemaldotearth committed:
fix: toast when token is created

Forgot to include the toast when an access token is created.

kemaldotearth committed:
chore: run prettier

kemaldotearth committed:
feat: plumbing new scoped access token endpoints

kemaldotearth committed:
feat: update some form bits

kemaldotearth committed:
chore: revert scoped api tests
